File: app/internal/publisher/template/python/crypto.py
"""Cryptographic utilities for API request signing"""
import base64
import hashlib
import hmac
import json
import uuid
from datetime import datetime, timezone
from typing import Any, Dict


def generate_timestamp() -> str:
    return datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z')


def generate_nonce() -> str:
    return str(uuid.uuid4())


def sign_request(
    payload: Dict[str, Any],
    secret: str,
    timestamp: str,
    nonce: str,
    method: str = 'POST',
    path: str = '',
    query: str = ''
) -> str:
    payload_json = json.dumps(payload, separators=(',', ':'))
    body_hash = hashlib.sha256(payload_json.encode('utf-8')).hexdigest()
    message = f"{method}\n{path}\n{query}\n{body_hash}\n{timestamp}\n{nonce}"
    signature = hmac.new(
        secret.encode('utf-8'),
        message.encode('utf-8'),
        hashlib.sha256
    ).digest()
    return base64.b64encode(signature).decode('utf-8')

File: app/internal/publisher/template/typescript/crypto.ts
import * as crypto from 'crypto';

export function generateTimestamp(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

export function generateNonce(): string {
  return crypto.randomUUID();
}

export function signRequest(
  payload: Record<string, unknown>,
  secret: string,
  timestamp: string,
  nonce: string,
  method: string = 'POST',
  path: string = '',
  query: string = '',
): string {
  const payloadJson = JSON.stringify(payload);
  const bodyHash = crypto.createHash('sha256').update(payloadJson, 'utf-8').digest('hex');
  const message = [method, path, query, bodyHash, timestamp, nonce].join('\n');
  const signature = crypto.createHmac('sha256', secret).update(message, 'utf-8').digest();
  return signature.toString('base64');
}
